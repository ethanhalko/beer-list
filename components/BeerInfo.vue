<script setup>
const props = defineProps({
  beer: {
    type: Object,
    required: true
  }
});

const readMore = ref(false);
const containsLactose = !!props.beer.ingredients.hops.find((hops) => hops.name.toLowerCase() === 'lactose');
const dryHopped = !!props.beer.ingredients.hops.find((hops) => hops.add?.toLowerCase() === 'dry hop');
</script>
<template>
  <div class="beer-card group flex flex-col" :class="{'max-h-64 sm:max-h-36': !readMore}" @click="readMore = !readMore">
    <div class="flex flex-row">
      <div class="card-image w-1/6" :class="{'sm:my-auto': readMore}">
        <img class="mx-auto max-h-24" :class="{'sm:max-h-36': readMore}" :src="beer.image_url" :alt="beer.description">
      </div>
      <div class="card-info w-5/6 flex flex-col">
        <div class="flex">
          <h1 class="beer-name sm:text-xl px-1 -ml-1" :class="{'bg-green-200': dryHopped}">{{ beer.name }}</h1>
          <Icon v-if="containsLactose" name="fa6-solid:cow" class="icon-lactose my-auto text-middle mx-1 w-6 text-red-600"></Icon>
        </div>
        <h2 class="text-md sm:my-auto">ABV: {{ beer.abv }} | IBU: {{ beer.ibu }}</h2>
        <h3 class="sm:text-lg italic">{{ beer.tagline }}</h3>
        <p v-if="readMore" class="description sm:text-lg pt-4 mt-auto">{{ beer.description }}</p>
      </div>
    </div>
    <p v-if="!readMore" class="text-sm italic mx-auto invisible group-hover:visible text-gray-400">Click To Expand</p>
  </div>
</template>
<style>
.beer-card {
  @apply my-4 p-4 rounded-md border hover:shadow-md transition-shadow duration-200 cursor-pointer;
}
</style>

