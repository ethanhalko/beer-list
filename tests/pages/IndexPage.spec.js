// @vitest-environment nuxt
import { describe, it, expect } from 'vitest';
import Index from '../../pages/index.vue';
import {  mountSuspended, registerEndpoint } from 'nuxt-vitest/utils';
import BeerInfo from '~/components/BeerInfo.vue';
import BeerListPayload from '../stubs/BeerListPayload.json';

registerEndpoint('https://api.punkapi.com/v2/beers?brewed_after=11-2012', () => BeerListPayload);

describe('BeerList', async () => {
  const wrapper = await mountSuspended(Index, {
    global: {
      stubs: {
        fas: () => ({
          template: '<span />'
        })
      }
    }
  });

  it('renders a vue component', () => {
    expect(wrapper.vm).toBeTruthy();
  });

  it('matches snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('filters centennial hops', () => {
    expect(wrapper.findAllComponents(BeerInfo).length).toBeLessThan(BeerListPayload.length);
  });

  it('orders by ABV desc', () => {
    const listValue = wrapper.getCurrentComponent().exposed?.beerList?.value;
    expect(listValue).toBeDefined();

    for (let i = 1; i < listValue.length; ++i) {
      expect(listValue[i].abv).toBeGreaterThan(listValue[i - 1].abv);
    }
  });
});
