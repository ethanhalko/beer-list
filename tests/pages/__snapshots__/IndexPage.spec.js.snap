// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BeerList > matches snapshot 1`] = `
"<div class=\\"flex flex-col mx-auto sm:w-1/2\\">
  <h1 class=\\"mx-auto font-bold text-gray-800 text-5xl my-10 drop-shadow-sm\\">Our Beers</h1>
  <h2 class=\\"mx-auto flex\\"><svg data-v-4fa0a2a1=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" xmlns:xlink=\\"http://www.w3.org/1999/xlink\\" aria-hidden=\\"true\\" role=\\"img\\" class=\\"my-auto text-middle mx-1 w-6 text-red-600\\" width=\\"1em\\" height=\\"1em\\" viewBox=\\"0 0 640 512\\">
      <path fill=\\"currentColor\\" d=\\"M96 224v192c0 17.7 14.3 32 32 32h32c17.7 0 32-14.3 32-32v-88.2c9.9 6.6 20.6 12 32 16.1V368c0 8.8 7.2 16 16 16s16-7.2 16-16v-16.9c5.3.6 10.6.9 16 .9s10.7-.3 16-.9V368c0 8.8 7.2 16 16 16s16-7.2 16-16v-24.2c11.4-4 22.1-9.4 32-16.1V416c0 17.7 14.3 32 32 32h32c17.7 0 32-14.3 32-32V256l32 32v49.5c0 9.5 2.8 18.7 8.1 26.6L530 427c8.8 13.1 23.5 21 39.3 21c22.5 0 41.9-15.9 46.3-38l20.3-101.6c2.6-13-.3-26.5-8-37.3l-3.9-5.5V184c0-13.3-10.7-24-24-24s-24 10.7-24 24v14.4l-52.9-74.1C496 86.5 452.4 64 405.9 64H144C77.7 64 24 117.7 24 184v54C9.4 249.8 0 267.8 0 288v17.6c0 8 6.4 14.4 14.4 14.4c31.8 0 57.6-25.8 57.6-57.6V184c0-24.3 12.1-45.8 30.5-58.9c-4.2 10.8-6.5 22.6-6.5 34.9v64zm464 112a16 16 0 1 1 32 0a16 16 0 1 1-32 0zM166.6 166.6c-4.2-4.2-6.6-10-6.6-16c0-12.5 10.1-22.6 22.6-22.6h178.8c12.5 0 22.6 10.1 22.6 22.6c0 6-2.4 11.8-6.6 16L354 190c-21.8 21.8-51.3 34-82 34s-60.2-12.2-81.9-33.9l-23.4-23.4z\\"></path>
    </svg> = Contains Lactose </h2>
  <h2 class=\\"mx-auto bg-green-300 p-1\\">Dry Hopped</h2>
  <div>
    <div class=\\"beer-card group flex flex-col max-h-64 sm:max-h-36\\">
      <div class=\\"flex flex-row\\">
        <div class=\\"card-image w-1/6\\"><img class=\\"mx-auto max-h-24\\" src=\\"https://images.punkapi.com/v2/keg.png\\" alt=\\"Japanese citrus fruit intensifies the sour nature of this German classic.\\"></div>
        <div class=\\"card-info w-5/6 flex flex-col\\">
          <div class=\\"flex\\">
            <h1 class=\\"beer-name sm:text-xl px-1 -ml-1\\">Berliner Weisse With Yuzu - B-Sides</h1>
            <!--v-if-->
          </div>
          <h2 class=\\"text-md sm:my-auto\\">ABV: 4.2 | IBU: 8</h2>
          <h3 class=\\"sm:text-lg italic\\">Japanese Citrus Berliner Weisse.</h3>
          <!--v-if-->
        </div>
      </div>
      <p class=\\"text-sm italic mx-auto invisible group-hover:visible text-gray-400\\">Click To Expand</p>
    </div>
  </div>
  <div>
    <div class=\\"beer-card group flex flex-col max-h-64 sm:max-h-36\\">
      <div class=\\"flex flex-row\\">
        <div class=\\"card-image w-1/6\\"><img class=\\"mx-auto max-h-24\\" src=\\"https://images.punkapi.com/v2/8.png\\" alt=\\"Fake is the new black. Fake is where it is at. Fake Art, fake brands, fake breasts, and fake lager. We want to play our part in the ugly fallout from the Lager Dream. Say hello to Fake Lager – a zesty, floral 21st century faux masterpiece with added BrewDog bitterness.\\"></div>
        <div class=\\"card-info w-5/6 flex flex-col\\">
          <div class=\\"flex\\">
            <h1 class=\\"beer-name sm:text-xl px-1 -ml-1\\">Fake Lager</h1>
            <!--v-if-->
          </div>
          <h2 class=\\"text-md sm:my-auto\\">ABV: 4.7 | IBU: 40</h2>
          <h3 class=\\"sm:text-lg italic\\">Bohemian Pilsner.</h3>
          <!--v-if-->
        </div>
      </div>
      <p class=\\"text-sm italic mx-auto invisible group-hover:visible text-gray-400\\">Click To Expand</p>
    </div>
  </div>
  <div>
    <div class=\\"beer-card group flex flex-col max-h-64 sm:max-h-36\\">
      <div class=\\"flex flex-row\\">
        <div class=\\"card-image w-1/6\\"><img class=\\"mx-auto max-h-24\\" src=\\"https://images.punkapi.com/v2/6.png\\" alt=\\"Re-brewed as a spring seasonal, this beer – which appeared originally as an Equity Punk shareholder creation – retains its trademark spicy, fruity edge. A perfect blend of Belgian Saison and US IPA, crushed peppercorns and heather honey are also added to produce a genuinely unique beer.\\"></div>
        <div class=\\"card-info w-5/6 flex flex-col\\">
          <div class=\\"flex\\">
            <h1 class=\\"beer-name sm:text-xl px-1 -ml-1\\">Electric India</h1>
            <!--v-if-->
          </div>
          <h2 class=\\"text-md sm:my-auto\\">ABV: 5.2 | IBU: 38</h2>
          <h3 class=\\"sm:text-lg italic\\">Vibrant Hoppy Saison.</h3>
          <!--v-if-->
        </div>
      </div>
      <p class=\\"text-sm italic mx-auto invisible group-hover:visible text-gray-400\\">Click To Expand</p>
    </div>
  </div>
  <div>
    <div class=\\"beer-card group flex flex-col max-h-64 sm:max-h-36\\">
      <div class=\\"flex flex-row\\">
        <div class=\\"card-image w-1/6\\"><img class=\\"mx-auto max-h-24\\" src=\\"https://images.punkapi.com/v2/12.png\\" alt=\\"Running the knife-edge between an India Pale Ale and a Stout, this particular style is one we truly love. Black IPAs are a great showcase for the skill of our brew team, balancing so many complex and twisting flavours in the same moment. The citrus, mango and pine from the hops – three of our all-time favourites – play off against the roasty dryness from the malt bill at each and every turn.\\"></div>
        <div class=\\"card-info w-5/6 flex flex-col\\">
          <div class=\\"flex\\">
            <h1 class=\\"beer-name sm:text-xl px-1 -ml-1 bg-green-200\\">Arcade Nation</h1>
            <!--v-if-->
          </div>
          <h2 class=\\"text-md sm:my-auto\\">ABV: 5.3 | IBU: 60</h2>
          <h3 class=\\"sm:text-lg italic\\">Seasonal Black IPA.</h3>
          <!--v-if-->
        </div>
      </div>
      <p class=\\"text-sm italic mx-auto invisible group-hover:visible text-gray-400\\">Click To Expand</p>
    </div>
  </div>
  <div>
    <div class=\\"beer-card group flex flex-col max-h-64 sm:max-h-36\\">
      <div class=\\"flex flex-row\\">
        <div class=\\"card-image w-1/6\\"><img class=\\"mx-auto max-h-24\\" src=\\"https://images.punkapi.com/v2/keg.png\\" alt=\\"The brainchild of our small batch brewer, George Woods. A dangerously drinkable milk sugar- infused Scotch Ale.\\"></div>
        <div class=\\"card-info w-5/6 flex flex-col\\">
          <div class=\\"flex\\">
            <h1 class=\\"beer-name sm:text-xl px-1 -ml-1\\">Misspent Youth</h1><svg data-v-4fa0a2a1=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" xmlns:xlink=\\"http://www.w3.org/1999/xlink\\" aria-hidden=\\"true\\" role=\\"img\\" class=\\"icon-lactose my-auto text-middle mx-1 w-6 text-red-600\\" width=\\"1em\\" height=\\"1em\\" viewBox=\\"0 0 640 512\\">
              <path fill=\\"currentColor\\" d=\\"M96 224v192c0 17.7 14.3 32 32 32h32c17.7 0 32-14.3 32-32v-88.2c9.9 6.6 20.6 12 32 16.1V368c0 8.8 7.2 16 16 16s16-7.2 16-16v-16.9c5.3.6 10.6.9 16 .9s10.7-.3 16-.9V368c0 8.8 7.2 16 16 16s16-7.2 16-16v-24.2c11.4-4 22.1-9.4 32-16.1V416c0 17.7 14.3 32 32 32h32c17.7 0 32-14.3 32-32V256l32 32v49.5c0 9.5 2.8 18.7 8.1 26.6L530 427c8.8 13.1 23.5 21 39.3 21c22.5 0 41.9-15.9 46.3-38l20.3-101.6c2.6-13-.3-26.5-8-37.3l-3.9-5.5V184c0-13.3-10.7-24-24-24s-24 10.7-24 24v14.4l-52.9-74.1C496 86.5 452.4 64 405.9 64H144C77.7 64 24 117.7 24 184v54C9.4 249.8 0 267.8 0 288v17.6c0 8 6.4 14.4 14.4 14.4c31.8 0 57.6-25.8 57.6-57.6V184c0-24.3 12.1-45.8 30.5-58.9c-4.2 10.8-6.5 22.6-6.5 34.9v64zm464 112a16 16 0 1 1 32 0a16 16 0 1 1-32 0zM166.6 166.6c-4.2-4.2-6.6-10-6.6-16c0-12.5 10.1-22.6 22.6-22.6h178.8c12.5 0 22.6 10.1 22.6 22.6c0 6-2.4 11.8-6.6 16L354 190c-21.8 21.8-51.3 34-82 34s-60.2-12.2-81.9-33.9l-23.4-23.4z\\"></path>
            </svg>
          </div>
          <h2 class=\\"text-md sm:my-auto\\">ABV: 7.3 | IBU: 30</h2>
          <h3 class=\\"sm:text-lg italic\\">Milk & Honey Scotch Ale.</h3>
          <!--v-if-->
        </div>
      </div>
      <p class=\\"text-sm italic mx-auto invisible group-hover:visible text-gray-400\\">Click To Expand</p>
    </div>
  </div>
  <div>
    <div class=\\"beer-card group flex flex-col max-h-64 sm:max-h-36\\">
      <div class=\\"flex flex-row\\">
        <div class=\\"card-image w-1/6\\"><img class=\\"mx-auto max-h-24\\" src=\\"https://images.punkapi.com/v2/19.png\\" alt=\\"We sent this beer to Norway where it was known as 'Hello, my name is Censored’. You can make up your own mind as to why. This brew was a red berry explosion, with a reisnous bitter edge layered with dry berry tartness.\\"></div>
        <div class=\\"card-info w-5/6 flex flex-col\\">
          <div class=\\"flex\\">
            <h1 class=\\"beer-name sm:text-xl px-1 -ml-1 bg-green-200\\">Hello My Name Is Mette-Marit</h1>
            <!--v-if-->
          </div>
          <h2 class=\\"text-md sm:my-auto\\">ABV: 8.2 | IBU: 70</h2>
          <h3 class=\\"sm:text-lg italic\\">Lingonberry Double IPA.</h3>
          <!--v-if-->
        </div>
      </div>
      <p class=\\"text-sm italic mx-auto invisible group-hover:visible text-gray-400\\">Click To Expand</p>
    </div>
  </div>
</div>"
`;
